util/stopper.go

```diff
@@ -17,39 +17,146 @@
+// to shut down. Once shutdown, each worker invokes SetStopped(). When
+// all workers have shutdown, the stopper is complete.
 type Stopper struct {
 	stopper chan struct{}
-	wg      sync.WaitGroup
+	state   int32
+	drain   sync.WaitGroup
+	stop    sync.WaitGroup
+	// mu protects the slice of Closers and adding/waiting on the
+	// stop WaitGroup.
+	mu      sync.Mutex
+	closers []Closer
 }
 
 // NewStopper returns an instance of Stopper. Count specifies how
-// many goroutines this stopper will stop.
-func NewStopper(count int) *Stopper {
-	s := &Stopper{
+// many workers this stopper will stop.
+func NewStopper() *Stopper {
+	return &Stopper{
 		stopper: make(chan struct{}),
+		state:   RUNNING,
+	}
+}
+
+func (s *Stopper) setState(newS int32) {
+	oldS := s.getState()
+	if !atomic.CompareAndSwapInt32(&s.state, oldS, newS) {
+		panic("concurrent access changed state")
+	}
+}
+
+func (s *Stopper) getState() int32 {
+	return atomic.LoadInt32(&s.state)
+}
+
+// AddWorker worker to the stopper.
+func (s *Stopper) AddWorker() {
+	// If we're already draining, we might already be waiting on the stop
+	// WaitGroup, and it's a data race to add to it now. We hold the lock
+	// while waiting, so obtaining the lock here prevents that race.
+	// This means that possibly we only add to the WaitGroup after the
+	// stopper has shut down, but Closers added when the stopper is not are
+	// closed immediately, so the worker will stop anyways.
+	//
+	// This will likely never happen outside of tests, where we sometimes
+	// create and stop components in rapid succession.
+	s.mu.Lock()
+	s.stop.Add(1)
+	s.mu.Unlock()
+}
+
+// AddCloser adds an object to close after the stopper has been stopped.
+func (s *Stopper) AddCloser(c Closer) {
+	// Hold the lock to make sure we don't cut through a half-way Stop().
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// If the state is not RUNNING, and we have the lock, then we should already
+	// done closing the stoppers in s.closers. The right thing to do is to just
+	// close it directly, and to never add it to the slice.
+	if s.getState() != RUNNING {
+		c.Close()
+		return
 	}
-	s.Add(count)
-	return s
+	// Otherwise, the closers are not yet being processed, so we can add it.
+	s.closers = append(s.closers, c)
 }
 
-// Add a new goroutine to the stopper.
-func (s *Stopper) Add(count int) {
-	s.wg.Add(count)
+// StartTask adds one to the count of tasks left to drain in the
+// system. Any worker which is a "first mover" when starting tasks
+// must call this method before starting work on a new task and must
+// subsequently invoke FinishTask() when the task is complete.
+// First movers include goroutines launched to do periodic work and
+// the kv/db.go gateway which accepts external client
+// requests.
+//
+// Returns true if the task can be launched or false to indicate the
+// system is currently draining and the task should be refused.
+func (s *Stopper) StartTask() bool {
+	if s.getState() == RUNNING {
+		s.drain.Add(1)
```

Comment:
I'm not seeing the point of the atomic operations for manipulating `Stopper.state`. And on the flip side it looks like it could be problematic. For example, between the `s.getState() == RUNNING` statement and the `s.drain.Add(1)` statement the stopper could close and finish.